proxy/lib/ketama.c

```diff
@@ -0,0 +1,233 @@
+
+        if ( midp == kt->total_buckets )
+            return kt->continuum[0].id; // if at the end, roll back to zeroth
+
+        midval = kt->continuum[midp].point;
+        midval1 = midp == 0 ? 0 : kt->continuum[midp-1].point;
+
+        if ( h <= midval && h > midval1 )
+            return kt->continuum[midp].id;
+
+        if ( midval < h )
+            lowp = midp + 1;
+        else
+            highp = midp - 1;
+
+        if ( lowp > highp )
+            return kt->continuum[0].id;
+    }
+}
+/* END FROM ketama.c */
+
+#define PARTS 3
+// stack = [pool, option]
+static int ketama_new(lua_State *L) {
+    lua_Integer bucket_size = DEFAULT_BUCKET_SIZE;
+    const char *parts[PARTS];
+    size_t partlens[PARTS];
+
+    // check for UA_TTABLE at 1
+    luaL_checktype(L, 1, LUA_TTABLE);
+    // get number of servers in pool.
+    // NOTE: rawlen skips metatable redirection. if we care; lua_len instead.
+    lua_Unsigned total = lua_rawlen(L, 1);
+    // check for optional input (set bucket_size)
+    int argc = lua_gettop(L);
+    if (argc > 1) {
+        // override default bucket_size if given
+        int success = 0;
+        bucket_size = lua_tointegerx(L, 2, &success);
+        if (!success) {
+            lua_pushfstring(L, "%s: option argument must be an integer", __func__);
+            lua_error(L);
+        }
+    }
+    // newuserdatauv() sized for pool*
+    size_t size = sizeof(ketama_t) + sizeof(cpoint) * (total * bucket_size);
+    ketama_t *kt = lua_newuserdatauv(L, size, 0);
+    // TODO: check *kt.
+    kt->total_buckets = bucket_size * total;
+
+    // loop over pool
+    unsigned int cont = 0;
+    lua_pushnil(L); // start the pool iterator
+    while (lua_next(L, 1) != 0) {
+        // key is -2, value is -1.
+        // value is another table. need to query it to get what we need for
+        // the hash.
+        // hash string is: hostname/ipaddr:port-repitition
+        // TODO: bother doing error checking?
+        lua_getfield(L, -1, "id");
+        lua_Integer id = lua_tointeger(L, -1);
+        lua_pop(L, 1);
+
+        // FIXME: we need to do the lua_pop after string assembly to be safe.
+        lua_getfield(L, -1, "hostname");
+        parts[0] = lua_tolstring(L, -1, &partlens[0]);
+        lua_pop(L, 1);
+        lua_getfield(L, -1, "addr");
+        parts[1] = lua_tolstring(L, -1, &partlens[1]);
+        lua_pop(L, 1);
+        lua_getfield(L, -1, "port");
+        parts[2] = lua_tolstring(L, -1, &partlens[2]);
+        lua_pop(L, 1);
+
+        size_t total = 0;
+        for (int x = 1; x < PARTS; x++) {
+            total += partlens[x];
+        }
+        // add extra space for the repititons.
+        char *buf = malloc(total + 16);
+
+        for (int k = 0; k < bucket_size / 4; k++) {
+            unsigned char digest[16];
+
+            // - create hashing string for ketama
+            snprintf(buf, total, "%s/%s:%s-%d", parts[0], parts[1], parts[2], k);
```

Comment:
As noted in my other client that other lib is for compatibility with EVCache, so I can't change something there. It is a completely loadable module so if someone wants something specific they can just do it. I haven't decided what I will use for the built-in multi-probe hashing, but it will likely be completely label based rather than hostname/ip/dns/etc for the same reasons you mention.